src/Security/MemberAuthenticator/ChangePasswordHandler.php

```diff
@@ -63,102 +66,152 @@ public function __construct($link, MemberAuthenticator $authenticator)
      * @return array|HTTPResponse
      */
     public function changepassword()
+    {
+        $ret = $this->processChangePasswordUrlVars();
+        if ($ret) {
+            return $ret;
+        }
+        return $this->createChangePasswordResponse();
+    }
+
+    /**
+     * Process the get URL variables for a change password request.
+     * If additional action is required, return a response for that action.
+     * Otherwise, return null which indicates the change password form should be presented.
+     */
+    protected function processChangePasswordUrlVars(): array|HTTPResponse|null
     {
         $request = $this->getRequest();
 
-        // Extract the member from the URL.
+        // Check if we're resetting via a token in URL i.e. reset password email link
         $member = null;
         if ($request->getVar('m') !== null) {
             $member = Member::get()->filter(['ID' => (int)$request->getVar('m')])->first();
         }
+
+        // If we have a token and a member, a user is trying to reset their password.
+        // We'll set a temp token and redirect back here for the next round of processing.
         $token = $request->getVar('t');
+        if ($token !== null && $member !== null) {
+            if (!$member->validateAutoLoginToken($token)) {
+                return $this->getInvalidTokenResponse();
+            }
+            // Redirect to current url, though with a temporary hash in the URL.
+            // This will ensure that that the member ID and token will not appear in browser history.
+            // Instead only a harmless temporary hash will appear in the browser history.
+            // We do this instead of setting the session value at this point because if
+            // cookie SameSite is set to Strict, then when clicking a password reset link via a
+            // webmail client the redirect will be treated as cross-origin request and value
+            // of the session cookie will not be accessible, so we have to set the session variable
+            // AFTER the redirect.
+            $autoLoginTempHash = $this->createAutoLoginTempHash();
+            $member->AutoLoginTempHash = $autoLoginTempHash;
+            $member->write();
+            $response = $this->redirect(Controller::join_links($this->link, '?th=' . $autoLoginTempHash));
+            return $response;
+        }
 
-        // Check whether we are merely changing password, or resetting.
-        if ($token !== null && $member && $member->validateAutoLoginToken($token)) {
-            $this->setSessionToken($member, $token);
+        // If a member or password reset token was included on its own, the URL was invalid.
+        if ($token !== null || $member !== null) {
+            return $this->getInvalidTokenResponse();
+        }
```

Comment:
This condition wasn't part of the original PRs and handles an edge case where only one of `?t` or `?m` is included in the URL (won't happen normally but can happen if someone copies a URL incorrectly for example)